docs/spec/slashing/tombstone.md

```diff
@@ -0,0 +1,74 @@
+
+Doing this tombstone system and getting rid of the slashing period tracking, will make the `slashing` module way simpler, especially because we can remove the hooks between the `stake` and `slashing` modules.
+
+> Note: The tombstone concept, only applies to byzantine faults reported over ABCI.  For slashable offenses tracked by the state machine (such as liveness faults), as there is not a delay between infraction and slashing, no slashing period tracking is needed. Also, a liveness bug probably isn't so egregious that it mandates force unbonding all delegations, and so the current jail system is adequate.
+
+
+## Further improvements / Related proposals:
+
+### Single slashing amount
+
+Another optimization that can be made is that if we assume that all ABCI faults for Tendermint consensus are slashed at the same level, we don't have to keep track of "max slash".  Once an ABCI fault happens once, we don't have to worry about comparing potential future ones to find the max.
+
+I believe current Tendermint ABCI faults are primarily:
+- Unjustified precommits (double signs)
+- Signing a precommit when you're in unbonding phase (can be used to trick light clients)
+
+Do we want to punish these two faults at different levels?  If not, we can enact the above change.  Note:  This change may make sense for current Tendermint consensus, but maybe not for a different consensus algorithm or future versions of Tendermint that may want to punish at different levels (for example, partial slashing).
+
+### Store infractions in state instead of iterating over unbonds/redelegations
+
+Currently, every time evidence of a new fault comes in, we currently iterate over all of the unbonds/redelegations away from a validator to see if the slash affects them or not.  If it does, we decrease the "balance" of the `ubd` or `red`.  However, as the number of unbonds or redelegations can be very high, this might be very expensive.  Instead, we can store evidences for all infractions that happened in the last `Unbonding Period` in state, and then whenever a ubd or red hits maturity, it can check if it needs to be slashed by checking it against the last infraction that happened before they started unbonding/redelegating away.  Because we only need to store the infractions from the last unbon  
```

Comment:
The problem is that - for redelegations - we need to change the amount of bonded stake of the target validator, and we need to do that at the time of the slash (otherwise they're not really "equally accountable").

Reply:
Also it looks like the last sentence is in progress?